Templating docs (#3239)
kyma-project/test-infra: development/tools/cmd/rendertemplates/README.md

## Overview

> The Render Templates is a tool that reads the configuration from the [`templates/config.yaml`](../../../../templates/config.yaml) file and data files from [`templates/data`](../../../../templates/data) directory to generate output files, such as Prow component jobs. While `config.yaml` file can hold configuration for output file, you can place such data within data files, which should hold configuration for related output files. Having separate files with grouped data is cleaner and easier to maintain than one huge config file.

Reviewer: ```suggestion
The Render Templates is a tool that reads the configuration from the [`templates/config.yaml`](../../../../templates/config.yaml) file and data files from the [`templates/data`](../../../../templates/data) directory to generate output files, such as Prow component jobs. While `config.yaml` file can hold configuration for an output file, you can place such data within the data files that hold configuration for related output files. Having separate files with grouped data is cleaner and easier to maintain than one huge config file.
```

> The Render Templates is a tool that reads the configuration from the [`templates/config.yaml`](../../../../templates/config.yaml) file and data files from the [`templates/data`](../../../../templates/data) directory to generate output files, such as Prow component jobs. While `config.yaml` file can hold configuration for an output file, you can place such data within the data files that hold configuration for related output files. Having separate files with grouped data is cleaner and easier to maintain than one huge config file.

Reviewer: Missed this one, sorry. 
```suggestion
The Render Templates is a tool that reads the configuration from the [`templates/config.yaml`](../../../../templates/config.yaml) file and data files from the [`templates/data`](../../../../templates/data) directory to generate output files, such as Prow component jobs. While the `config.yaml` file can hold configuration for an output file, you can place such data within the data files that hold configuration for related output files. Having separate files with grouped data is cleaner and easier to maintain than one huge config file.
```

> The `config.yaml` file and `data files` specifies the following for the Render Templates:

Reviewer: `data files` should not be formatted as code unless it's a name of a file/directory.
```suggestion
The `config.yaml` file and the data files specify the following for the Render Templates:
```

> - Templates it must use to generate the output files
> - The name and location of the output files
> - Values it must use to generate the output files

Reviewer: Since you've added one more thing to the preceding line, we now refer to a plural, so `it` -> `they`/`them`. Unless it refers to the Render Templates, then it should have been plural from the start. 
```suggestion
- Templates they must use to generate the output files
- The name and location of the output files
- Values they must use to generate the output files
```

Author: Render Templates is a tool, that's why I used "it". It (Render Templates) must use templates to generate the output files.

Reviewer: I wasn't sure if you wanted to refer to it as a tool or as templates, as you used both singular and plural verbs throughout the document. That convinces me. Unfortunately, that means that we need to revert some of my corrections 
![image](https://user-images.githubusercontent.com/53601578/107653334-cb296200-6c81-11eb-9783-339cb65d925f.png)

### Data for templates 

> The Render Template passes data used to generate files from templates in the **$.Values** and **$.Global** variables. The values of these variables are created from the `config.yaml` file and data files stored in `templates/data` directory. The **Global** variable holds data from the **global** key. The Render Template generates the values of the **Values** variable from scratch for each Prow job. **Values** variable is generated by merging ConfigSets. Each ConfigSet is a map defined in one of three possible places:

Reviewer: I notices it says `Render Templates` everywhere else. Is `Render Template` a typo here? 

```suggestion
The Render Templates pass data used to generate files from templates in the **$.Values** and **$.Global** variables. The values of these variables are created from the `config.yaml` file and data files stored in the `templates/data` directory. The **Global** variable holds data from the **global** key. The Render Templates generate the values of the **Values** variable from scratch for each Prow job. **Values** variable is generated by merging ConfigSets. Each ConfigSet is a map defined in one of three possible places:
```

Or is it a collection of multiple Render Templates and you describe an exemplary one here? 
In that case, you should go with: 
```A Render Template passes data used to generate files from templates in the **$.Values** and **$.Global** variables. The values of these variables are created from the `config.yaml` file and data files stored in the `templates/data` directory. The **Global** variable holds data from the **global** key. The Render Template generates the values of the **Values** variable from scratch for each Prow job. **Values** variable is generated by merging ConfigSets. Each ConfigSet is a map defined in one of three possible places:```

> The Render Templates pass data used to generate files from templates in the **$.Values** and **$.Global** variables. The values of these variables are created from the `config.yaml` file and data files stored in the `templates/data` directory. The **Global** variable holds data from the **global** key. The Render Templates generate the values of the **Values** variable from scratch for each Prow job. **Values** variable is generated by merging ConfigSets. Each ConfigSet is a map defined in one of three possible places:

Reviewer: If you mean it as a tool here, and not as Templates, then it should be singular here. You used both singular and plural verbs throughout the document so I wasn't sure.
I should have clarified this before correction.
```suggestion
The Render Templates passes data used to generate files from templates in the **$.Values** and **$.Global** variables. The values of these variables are created from the `config.yaml` file and data files stored in the `templates/data` directory. The **Global** variable holds data from the **global** key. The Render Templates generates the values of the **Values** variable from scratch for each Prow job. **Values** variable is generated by merging ConfigSets. Each ConfigSet is a map defined in one of three possible places:
```

   ConfigSets defined in **globalSets** hold data used to generate multiple files. A good example of such usage is the `image_bootstrap` global ConfigSet which defines a bootstrap image to use in Prow jobs. 
 
 

> - **Local ConfigSets** defined under the **localSets** parameter for each **to** key in the `config.yaml` file or in data files in `templates/data` directory:

Reviewer: ```suggestion
- **Local ConfigSets** defined under the **localSets** parameter for each **to** key in the `config.yaml` file or in data files in the `templates/data` directory:
```

          name: "pre-test-infra-bootstrap"
          run_if_changed: "^prow/images/bootstrap/"
```

> Render Templates builds the **Values** variable by merging ConfigSets from **globalSets** first. If the job inherits the `default` ConfigSet from **globalSets**, it's merged first and all other ConfigSets from **globalSets** are merged after. Then, Rener Templates merge ConfigSets from **localSets**. Again, if the job inherits the `default` ConfigSet from **localSets**, it's merged first and then all the other ConfigSets from **localSets** are merged. ConfigSets other than default are merged in any order during the **globalSets** and **localSets** phases. ConfigSets from **jobConfig** are merged as the last ones. Existing keys in the **Values** variable are overwritten by values from the merged ConfigSets.

Reviewer: Render Templates is plural 
```suggestion
Render Templates build the **Values** variable by merging ConfigSets from **globalSets** first. If the job inherits the `default` ConfigSet from **globalSets**, it is merged first and all other ConfigSets from **globalSets** are merged afterwards. Then, Render Templates merge ConfigSets from **localSets**. Again, if the job inherits the `default` ConfigSet from **localSets**, it's merged first and then all the other ConfigSets from **localSets** are merged. ConfigSets other than default are merged in any order during the **globalSets** and **localSets** phases. ConfigSets from **jobConfig** are merged as the last ones. Existing keys in the **Values** variable are overwritten by values from the merged ConfigSets.
```

> Render Templates build the **Values** variable by merging ConfigSets from **globalSets** first. If the job inherits the `default` ConfigSet from **globalSets**, it is merged first and all other ConfigSets from **globalSets** are merged afterwards. Then, Render Templates merge ConfigSets from **localSets**. Again, if the job inherits the `default` ConfigSet from **localSets**, it's merged first and then all the other ConfigSets from **localSets** are merged. ConfigSets other than default are merged in any order during the **globalSets** and **localSets** phases. ConfigSets from **jobConfig** are merged as the last ones. Existing keys in the **Values** variable are overwritten by values from the merged ConfigSets.

Reviewer: ```suggestion
The Render Templates builds the **Values** variable by merging ConfigSets from **globalSets** first. If the job inherits the `default` ConfigSet from **globalSets**, it is merged first and all other ConfigSets from **globalSets** are merged afterwards. Then, the Render Templates merges ConfigSets from **localSets**. Again, if the job inherits the `default` ConfigSet from **localSets**, it's merged first and then all the other ConfigSets from **localSets** are merged. ConfigSets other than default are merged in any order during the **globalSets** and **localSets** phases. ConfigSets from **jobConfig** are merged as the last ones. Existing keys in the **Values** variable are overwritten by values from the merged ConfigSets.
```
